backend/flaskr/database_utils.py

```diff
@@ -2,57 +2,85 @@
+        :return: None
+        """
 
-    # Try to connect to the database
-    try:
-        return mysql.connect()
-    except (AttributeError, OperationalError):
-        return None
+    @classmethod
+    def setup(cls, app, **kwargs):
+        """
+        Initializes MySQL at the start of Flask app. Should be run once per app instance.
+        :param app: Flask app instance
+        :param kwargs: Allows for nonstandard MySQL parameters
+        :return: None
+        """
+        # Initialize mysql object and connection pool
+        cls.mysql = MySQL()
+        cls.connections = []
+        # Do a standard configuration
+        app.config.from_object(Config())
+        # Modify configuration based on arguments
+        for arg, value in kwargs.items():
+            if arg == "socket":
+                app.config["MYSQL_DATABASE_HOST"] = value[0]
+                app.config["MYSQL_DATABASE_PORT"] = value[1]
+            elif arg == "credentials":
+                app.config["MYSQL_DATABASE_USER"] = value[0]
+                app.config["MYSQL_DATABASE_PASSWORD"] = value[1]
+            elif arg == "database":
+                app.config["MYSQL_DATABASE_DB"] = value
+            elif arg == "charset":
+                app.config["MYSQL_DATABASE_CHARSET"] = value
+        # Initialize app and make one connection
+        cls.mysql.init_app(app)
+        cls.return_connection(cls.get_connection())
 
-def execute_auth_query(app, query):
-    """
-    This function executes a query on the MySQL server.
-    :param app: Flask app instance
-    :param query: The query to be executed
-    :return: The first result of the query
-    """
-    conn = get_db_connection(app, database="authdb")
-    if conn is None:
-        print("Could not connect to database!")
-        return None
-    cursor = conn.cursor()
-    cursor.execute(query)
-    result = cursor.fetchone()
-    conn.commit()
-    conn.close()
-    return result
+    @classmethod
+    def get_connection(cls):
+        """
+        Gets a connection from the pool. Creates a new connection if none are available.
+        :return: Database connection object
+        """
+        if cls.connections:
+            return cls.connections.pop()
+        try:
+            return cls.mysql.connect()
+        except (AttributeError, OperationalError):
+            print("Couldn't get connection!")
+            return None
+
+    @classmethod
+    def return_connection(cls, conn):
+        """
+        Returns a database connection to the connection pool.
+        :param conn: Database connection object
+        :return: None
+        """
+        cls.connections.append(conn)
+
+    def execute_query(self, query):
+        """
+        Gets a connection, executes a query, and returns the connection.
+        :param query: SQL query to be executed
+        :return: First result of the query
+        """
+        conn = self.get_connection()
+        cursor = conn.cursor()
```

Comment:
We get a null pointer exception here if the database connection fails. I think it would be better to check if `conn` is null here first and handle the error ourselves.

Reply:
We now return -1 if a db connection cannot be established. The login and registration functions behave as expected (registration page does not try to authenticate a user after a failed registration, and login fails if no connection can be made).